software/due-date-changes.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+
+import argparse
+import json
+import re
+import sys
+import time
+
+from datetime import datetime
+from jira import JIRA
+
+
+def dump(obj):
+    for attr in dir(obj):
+        print("obj.%s = %r" % (attr, getattr(obj, attr)))
+
+
+def print_help(stream=sys.stderr):
+    help_msg = """Usage: {0} [--projects] LIST,OF,PROJECTS
+"""
+    stream.write(help_msg.format(sys.argv[0]))
+
+
+def parse_args():
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument("-projects", 
+        help="Comma-separated list of project names (ie. HTCONDOR,SOFTWARE)",
+        default="HTCONDOR,SOFTWARE")
+    parser.add_argument("-o", 
+        help="Output file",
+        default=None)
+    args = parser.parse_args()
+
+    projects = args.projects
+    output_file = args.o
+
+    return {
+        "projects": projects,
+        "output_file": output_file
+    }
+
+
+def main():
+
+    # Parse arguments. Assume any error handling happens in parse_args()
+    try:
+        args = parse_args()
+    except Exception as err:
+        print(f"Failed to parse arguments: {err}", file=sys.stderr)
+
+    projects = args["projects"]
+    output_file = args["output_file"]
+
+    # Connect to Jira
+    options = {"server": "https://opensciencegrid.atlassian.net"}
+    jira = JIRA(options)
+
+    # Output CSV headers
+    output_data = "Issue key,Assignee,Original Due Date,Current Due Date,Number of Due Date Changes\n"
+
+    # Now iterate over the projects and list their issues
+    for project in projects.split(","):
+        issues = jira.search_issues(f"project = {project} AND status != Backlog", expand="changelog", maxResults=False)
+        #issues = jira.search_issues(f"key = HTCONDOR-344", expand="changelog", maxResults=False)
+        for issue in issues:
+            duedate_current = issue.fields.duedate
+            duedate_original = issue.fields.duedate
+            duedate_changes = 0
+            issue_changelog = issue.changelog.histories
+            # Iterate over the changelog in reverse so we get the oldest changes first
+            for change in issue_changelog[::-1]:
+                for item in change.items:
+                    if item.field == "duedate":
+                        if duedate_changes == 0:
+                            duedate_original = getattr(item, "from")
+                        duedate_changes += 1
+                        duedate_current = item.to
+            output_data += f"{issue.key},{issue.fields.assignee},{duedate_original},{duedate_current},{duedate_changes}\n"
```

Comment:
Don't feel compelled to re-write it, but anyway just my two cents here -

It seems to me everything inside this `for issue in issues:` block might go nicely into a separate function, eg `def issue_info(issue):`, that returns this f-string you append at the end.

So then something like

```
        for issue in issues:
            output_data += issue_info(issue)
```

In particular that would make it more clear that the variables here (eg, `duedate_*`) are only relevant to and have an intended scope for this block (ie, for a single issue), rather than across all issues.  (Because currently their scope is the entire `main` function.)

(I admit i had to spend some time thinking about it to figure out if they were tracking something for a single issue or across issues.)

Reply:
I'm going to leave this as-is, it already works the way we want.